server/s3_put_object_action.cc

```diff
@@ -841,6 +834,119 @@ void S3PutObjectAction::add_object_oid_to_probable_dead_oid_list_failed() {
+        bucket_metadata->get_objects_version_list_index_layout().oid,
+        old_object_metadata->get_version_key_in_index(),
+        false /* force_delete */));
+    probable_oid_list[old_oid_rec_key] = old_probable_del_rec->to_json();
+    probable_del_rec_list.push_back(std::move(old_probable_del_rec));
+    old_obj_oids.push_back(old_object_oid);
+    old_obj_pvids.push_back(old_object_metadata->get_pvid());
+    old_obj_layout_ids.push_back(old_layout_id);
+  }
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3PutObjectAction::fetch_null_version_object_info() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  assert(!null_object_version_id.empty());
+  if (!null_object_metadata) {
+    null_object_metadata = object_metadata_factory->create_object_metadata_obj(
+        request, obj_list_idx_lo, obj_version_list_idx_lo);
+  }
+  null_object_metadata->set_version_id(null_object_version_id);
+  null_object_metadata->load(
+      std::bind(&S3PutObjectAction::fetch_null_version_object_info_success,
+                this),
+      std::bind(&S3PutObjectAction::fetch_null_version_object_info_failed,
+                this));
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+void S3PutObjectAction::fetch_null_version_object_info_success() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  auto omds = null_object_metadata->get_state();
+  if (omds == S3ObjectMetadataState::present) {
+    s3_log(S3_LOG_DEBUG, request_id, "S3ObjectMetadataState::present\n");
+    old_object_oid = null_object_metadata->get_oid();
+    old_oid_str = S3M0Uint128Helper::to_string(old_object_oid);
+    old_layout_id = null_object_metadata->get_layout_id();
+    number_of_parts = null_object_metadata->get_number_of_parts();
+    null_version_object_initialized = true;
+    if (old_object_oid.u_hi || old_object_oid.u_lo) {
+      _add_object_oid_to_probable_dead_oid_list(null_object_metadata);
+    }
+    update_global_probable_dead_object_list_index();
+  } else {
+    s3_log(S3_LOG_DEBUG, request_id,
+           "Unexpected Metadata state %d in success handler\n", (int)omds);
+    s3_put_action_state = S3PutObjectActionState::validationFailed;
+    set_s3_error("InternalError");
+    send_response_to_s3_client();
+  }
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3PutObjectAction::fetch_null_version_object_info_failed() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  const auto omds = null_object_metadata->get_state();
+  if (omds == S3ObjectMetadataState::missing) {
+    s3_log(
+        S3_LOG_DEBUG, request_id,
+        "Object with null version not found. Probable deleted explicitly.\n");
+    // Even though null version object is not present,
+    // update the global probable delete list, which will have current object's
+    // probable_del_record
+    update_global_probable_dead_object_list_index();
```

Comment:
Don't we need to update this record in all failure cases here?

Reply:
If there is a failure other than `S3ObjectMetadataState::missing` than that means something is wrong and we don't want procced and report a failure to the client.